lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -727,32 +740,44 @@ OpFoldResult AsUIntPrimOp::fold(ArrayRef<Attribute> operands) {
   if (input().getType() == getType())
     return input();
 
+  if (!operands[0])
+    return {};
+
   // Be careful to only fold the cast into the constant if the size is known.
   // Otherwise width inference may produce differently-sized constants if the
   // sign changes.
-  if (auto attr = operands[0].dyn_cast_or_null<IntegerAttr>())
+  if (auto attr = operands[0].dyn_cast<IntegerAttr>())
     if (getType().hasWidth())
       return getIntAttr(getType(), attr.getValue());
+
+  // Constant clocks and resets are bool attributes.
+  if (auto attr = operands[0].dyn_cast<BoolAttr>())
+    return getIntAttr(getType(), APInt(/*bitWidth*/ 1, attr.getValue()));
+
   return {};
 }
 
 OpFoldResult AsAsyncResetPrimOp::fold(ArrayRef<Attribute> operands) {
-  // TODO: Implement constants of asyncreset type.
-
   // No effect.
   if (input().getType() == getType())
     return input();
 
+  // Constant fold.
+  if (auto attr = operands[0].dyn_cast_or_null<IntegerAttr>())
+    return BoolAttr::get(getContext(), attr.getValue().getBoolValue());
+
   return {};
 }
 
 OpFoldResult AsClockPrimOp::fold(ArrayRef<Attribute> operands) {
-  // TODO: Implement constants of clock type.
-
   // No effect.
   if (input().getType() == getType())
     return input();
```

Comment:
Not related to your patch, but it would be better to do something like:
```
 if (same type)
   return operands[0] ? OpFoldResult(operands[0]) : OpFoldResult(input());
```

Because then symbolic constants will get propagated through the fold call.